README.md
# README

An abstract class which you may use when creating your custom modules instead of using standard prestashop Module class.

It implements some features that allows you to create modules in a more schematic way without unnecessarily repeating the code.

I plan to add more options in the future including simplification for creating controllers, forms, tables etc.

## Requirements

- PHP >= 5.2.4
- Prestashop 1.5 - 1.7

## Installation

### Composer (recommended)

```bash
$ composer require presta-expert/module-helper
```

## Basic usage

```php
<?php
/**
 * Checking if module is executed in prestashop context
 */
if (!defined('_PS_VERSION_')) {
    exit;
}

/**
 * Loading composer autoload
 */
require __DIR__ . '/vendor/autoload.php';

/**
 * Your module description
 */
class MyCustomModule extends \PrestaExpert\Helper\AbstractModule
{
    /**
     * Setting your module details
     */
    public function __construct()
    {
        $this->name    = 'mycustommodule';
        $this->version = '1.5.2';
        $this->author  = 'Johnny Sins';
        $this->tab     = 'analytics_stats';

        parent::__construct();

        $this->displayName = $this->l('My custom module');
        $this->description = $this->l('My custom module full description');
    }

    /**
     * {@inheritdoc}
     * 
     * Below configuration values will be automatically 
     * added on module installation
     *
     * Also below values will be automatically deleted
     * on module uninstallation
     */
    public function getConfiguration()
    {
        return array(
            'MY_CUSTOM_CONFIG_1' => 'Lorem',
            'MY_CUSTOM_CONFIG_2' => 'Lorem ipsum',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below hooks will be automatically registered on 
     * module installation without manually implementing 
     * registerHook
     * 
     * All you have to do is implement your hook methods
     * public function hookDisplayHeader(array $params)
     */
    public function getHooks()
    {
        return array(
            'displayHeader',
            'displayBackOfficeHeader',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below queries will be automatically executed on 
     * module installation
     */
    public function getInstallSql()
    {
        return array(
            'CREATE TABLE `custom_table` (
                 id int,
                 value1 varchar(255),
                 value2 varchar(255)
             );',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below queries will be automatically executed on 
     * module uninstallation
     */
    public function getUninstallSql()
    {
        return array(
            'DROP TABLE `custom_table`',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below tabs will be automatically added on module
     * installation
     * 
     * Also below tabs will be automatically deleted
     * on module uninstall
     *
     * Keep in mind to keep the proper structure of 
     * the array
     */
    public function getTabs()
    {
        return array(
            array(
                // Parent tab name (or null if your tab should be parent)
                'parent' => 'AdminParentOrders', 
    
                // Your tab class name
                'class' => 'AdminModule',
    
                // Display name of tab (will be automatically used in translation $this->l())
                'name' => 'Doing something',
            ),
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below media will be automatically added in 
     * back office using displayBackOfficeHeader
     */
    public function getBackOfficeMedia()
    {
        return array(
            '/modules/mycustommodule/views/js/backoffice.js',
            '/modules/mycustommodule/views/css/backoffice.css',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * Below media will be automatically added in 
     * front office using displayHeader
     */
    public function getFrontOfficeMedia()
    {
        return array(
            '/modules/mycustommodule/views/js/frontoffice.js',
            '/modules/mycustommodule/views/css/frontoffice.css',
        );
    }

    /**
     * {@inheritdoc}
     * 
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader when your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     * 
     * @return  void
     */
    public function hookDisplayBackOfficeHeader()
    {
        parent::hookDisplayBackOfficeHeader();
    
        // Your hook placeholder to keep getBackOfficeMedia() working
    }

    /**
     * {@inheritdoc}
     * 
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader when your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @return  void
     */
    public function hookDisplayHeader()
    {
        parent::hookDisplayHeader();
        
        // Your hook placeholder to keep getFrontOfficeMedia() working
    }

}
```

## Authors

- [Presta.Expert](https://presta.expert) Team

## License

The files in this archive are released under the [MIT LICENSE](LICENSE).


CHANGELOG.md
# CHANGELOG

## 1.0.5 (2023-02-21)

- Rebranding to PrestaExpert

## 1.0.4 (2020-03-30)

- Changed default module author because it looks ugly at prestashop 1.7

## 1.0.3 (2020-03-30)

- Added possibility to add hidden tab by setting "parent" as negative value i.e -1

## 1.0.2 (2020-03-29)

- Fixed release

## 1.0.1 (2020-03-28)

- Fixed typo which was making getTabs() unusable

## 1.0.0 (2020-03-20)

- First stable release


source/AbstractModule.php
<?php
/**
 * Namespace declaration
 */
namespace PrestaExpert\Helper;

/**
 * Checking if module is executed in prestashop context
 */
if (!defined('_PS_VERSION_')) {
    exit;
}

/**
 * Used namespaces
 */
use Db;
use Tab;
use Shop;
use Module;
use Configuration;

/**
 * An abstract class which makes creating modules
 * faster and without repeating the code
 *
 * Your module class should extend this class to
 * get additional features
 *
 * IMPORTANT NOTE:
 * Keep in mind to execute parent::hookDisplayHeader
 * and parent::hookDisplayBackOfficeHeader if your
 * module uses it to keep getFrontOfficeMedia() and
 * getBackOfficeMedia() working.
 *
 * @author      Presta.Expert Team <support@presta.expert>
 * @copyright   Presta.Expert Team <support@presta.expert>
 * @version     1.0.5
 */
abstract class AbstractModule extends Module
{
    /**
     * Module internal name
     *
     * @var string
     */
    public $name;

    /**
     * Module name which gonna be displayed
     *
     * @var string
     */
    public $displayName;

    /**
     * Module description which gonna be displayed
     *
     * @var string
     */
    public $description;

    /**
     * Module version
     *
     * @var string
     */
    public $version = '1.0.0';

    /**
     * Module author
     *
     * @var string
     */
    public $author = 'support@presta.expert';

    /**
     * Module tab / category
     *
     * administration
     * advertising_marketing
     * analytics_stats
     * billing_invoicing
     * checkout
     * content_management
     * dashboard
     * emailing
     * export
     * front_office_features
     * i18n_localization
     * market_place
     * merchandizing
     * migration_tools
     * mobile
     * others
     * payments_gateways
     * payment_security
     * pricing_promotion
     * quick_bulk_update
     * search_filter
     * seo
     * shipping_logistics
     * slideshows
     * smart_shopping
     * social_networks
     *
     * @var string
     */
    public $tab;

    /**
     * Should we create an instance of module on module list?
     *
     * @var integer
     */
    public $need_instance = 0;

    /**
     * Should our module use twitter bootstrap?
     *
     * @var integer
     */
    public $bootstrap = 1;

    /**
     * Supported versions of prestashop
     *
     * @var array
     */
    public $ps_versions_compliancy = array(
        'min' => '1.5',
        'max' => '1.7'
    );

    /**
     * Returns an array of configuration keys
     *
     * Returned keys will be automatically added
     * and removed on installing / uninstalling
     *
     * @return  array
     */
    public function getConfiguration()
    {
        return array();
    }

    /**
     * Returns an array of hooks to register
     *
     * Returned hooks will be automatically
     * added on installing our module
     *
     * @return  array
     */
    public function getHooks()
    {
        return array();
    }

    /**
     * Returns an array of additional sql queries
     * which will be executed on install
     *
     * You can place for example CREATE TABLE
     * queries here
     *
     * @return  array
     */
    public function getInstallSql()
    {
        return array();
    }

    /**
     * Returns an array of additional sql queries
     * which will be executed on uninstall
     *
     * You can place for example DROP TABLE here
     *
     * @return  array
     */
    public function getUninstallSql()
    {
        return array();
    }

    /**
     * Returns an array of media (css, js) which
     * should be added in back office
     *
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader if your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @return  array
     */
    public function getBackOfficeMedia()
    {
        return array();
    }

    /**
     * Returns an array of media (css, js) which
     * should be added in front office
     *
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader if your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @return  array
     */
    public function getFrontOfficeMedia()
    {
        return array();
    }

    /**
     * Returns an array of tabs which module should
     * add on installation
     *
     * return array(
     *     array(
     *         'parent' => 'AdminParentOrders',
     *         'class'  => 'AdminModule',
     *         'name'   => 'Doing something',
     *     ),
     * );
     *
     * @return  array
     */
    public function getTabs()
    {
        return array();
    }

    /**
     * {@inheritDoc}
     */
    public function install()
    {
        // Setting context for multi-store purposes
        if (Shop::isFeatureActive()) {
            Shop::setContext(Shop::CONTEXT_ALL);
        }

        // Calling parent::install();
        if (!parent::install()) {
            return false;
        }

        // Loop over each of configuration value
        foreach ($this->getConfiguration() as $key => $defaultValue) {

            // Trying to create configuration value
            if (!Configuration::updateValue($key, $defaultValue)) {
                return false;
            }
        }

        // Loop over each of hook
        foreach ($this->getHooks() as $hook) {

            // Trying to register hook
            if (!$this->registerHook($hook)) {
                return false;
            }
        }

        // Loop over each of sql
        foreach ($this->getInstallSql() as $sql) {

            // Trying to execute SQL query
            if (!Db::getInstance(_PS_USE_SQL_SLAVE_)->execute($sql)) {
                return false;
            }
        }

        // Loop over each of tab
        foreach ($this->getTabs() as $tabArray) {

            // Creating tab
            $tab = new Tab();
            $tab->id_parent = (int)Tab::getIdFromClassName($tabArray['parent']);
            $tab->class_name = $tabArray['class'];
            $tab->active = 1;
            $tab->module = $this->name;

            // Creating hidden tab
            if ($tabArray['parent'] < 0) {
                $tab->id_parent = -1;
            }

            // Creating translations
            foreach ($this->context->controller->getLanguages() as $lang) {

                // Adding translation for each lang
                $tab->name[(int)$lang['id_lang']] = $this->l($tabArray['name']);
            }

            // Trying to add new tab
            if (!$tab->add()) {
                return false;
            }
        }

        // Registering displayHeader hook if we have at least one media defined
        // No matter if user registered displayHeader earlier it will work anyways (tested)
        // All you have to remember it to execute parent::hookDisplayHeader in your module
        if ($this->getFrontOfficeMedia()) {

            // Registering displayHeader hook
            $this->registerHook('displayHeader');
        }

        // Registering displayBackOfficeHeader hook if we have at least one media defined
        // No matter if user registered displayBackOfficeHeader earlier it will work anyways (tested)
        // All you have to remember it to execute parent::hookDisplayBackOfficeHeader in your module
        if ($this->getBackOfficeMedia()) {

            // Registering displayBackOfficeHeader hook
            $this->registerHook('displayBackOfficeHeader');
        }

        // Returning boolean true
        return true;
    }

    /**
     * {@inheritDoc}
     */
    public function uninstall()
    {
        // Calling parent::uninstall();
        if (!parent::uninstall()) {
            return false;
        }

        // Loop over each of configuration value
        foreach (array_keys($this->getConfiguration()) as $key) {

            // Trying to remove configuration value
            if (!Configuration::deleteByName($key)) {
                return false;
            }
        }

        // Loop over each of sql
        foreach ($this->getUninstallSql() as $sql) {

            // Trying to execute SQL query
            if (!Db::getInstance(_PS_USE_SQL_SLAVE_)->execute($sql)) {
                return false;
            }
        }

        // Loop over each of tab
        foreach ($this->getTabs() as $tabArray) {

            // Creating tab
            $tab = new Tab((int)Tab::getIdFromClassName($tabArray['class']));

            // Trying to remove tab
            if (!$tab->delete()) {
                return false;
            }
        }

        // Returning true
        return true;
    }

    /**
     * Adds custom media (css & js) in back office
     * using displayBackOfficeHeader hook
     *
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader if your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @return  void
     */
    public function hookDisplayBackOfficeHeader()
    {
        return $this->addMedia($this->getBackOfficeMedia());
    }

    /**
     * Adds custom media (css & js) in front office
     * using displayHeader hook
     *
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader if your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @return  void
     */
    public function hookDisplayHeader()
    {
        return $this->addMedia($this->getFrontOfficeMedia());
    }

    /**
     * Adds media using controller addCSS or addJS
     * method based on extension
     *
     * It adds CSS in media type = all however you
     * can change it by extending this method
     *
     * IMPORTANT NOTE:
     * Keep in mind to execute parent::hookDisplayHeader
     * and parent::hookDisplayBackOfficeHeader when your
     * module uses it to keep getFrontOfficeMedia() and
     * getBackOfficeMedia() working.
     *
     * @param   array   $mediaArray
     * @return  void
     */
    protected function addMedia(array $mediaArray)
    {
        // Loop over each of element in array
        foreach ($mediaArray as $media) {

            // Exploding string using dot
            $explode = explode('.', $media);

            // Getting file extension
            $extension = mb_strtolower(end($explode));

            // Adding correct media depends on extension
            // I'm not sure if sass, scss, less works correctly but i've implemented them
            if (in_array($extension, ['css', 'sass', 'scss', 'less'])) {
                $this->context->controller->addCSS($media);
            }
            elseif ($extension == 'js') {
                $this->context->controller->addJS($media);
            }
        }
    }
}
